Lib/unittest/mock.py

```diff
@@ -467,10 +467,19 @@ def attach_mock(self, mock, attribute):
         Attach a mock as an attribute of this one, replacing its name and
         parent. Calls to the attached mock will be recorded in the
         `method_calls` and `mock_calls` attributes of this one."""
-        mock._mock_parent = None
-        mock._mock_new_parent = None
-        mock._mock_name = ''
-        mock._mock_new_name = None
+
+        # When autospecced object is passed to attach_mock then clear
+        # name and parent of the mock attribute which holds the actual
+        # mock object.
+        if isinstance(mock, FunctionTypes) and hasattr(mock, 'mock'):
```

Comment:
If this is the way that we know an autospecced object is passed in, it feels like it should be abstracted out into a helper function with a sensible name, and used wherever else this pattern has cropped up.

Reply:
I moved it to `_extract_mock` to avoid code duplication and added a code comment related to autospecced functions.